src/toolchain-conventions.adoc

```diff
@@ -203,6 +203,89 @@ RISC-V psABI has a contain sections to describe the code model:
+known as *negative directive*. A **positive directive**, on the other hand, enables
+an additional tuning feature in the associated tuning model. For example,
+`-mtune=sifive-x280:single-element-vec-fp64` represents a set of performance
+characteristics that is a union of that from `sifive-x280` and `single-element-vec-fp64`,
+which assumes certain FP64 vector operations to produce a single element per cycle.
+The `single-element-vec-fp64` in this case is a positive directive.
+
+It is in our best interests to align the directives that are supported by multiple compilers.
+However, it is understandable that the set of supported directives might vary from
+one compiler to another. Therefore, unrecognized directives are ignored for compatibility
+reasons, and the compiler may optionally emit a warning.
+
+The tuning features enabled by positive directives in a tune feature string cannot
+overlap with tuning features disabled by negative directives in the same string.
+This mandate applies to explicitly induced tuning features as well as tuning
+features that are implied _implicitly_. For example, `short-forward-branch-i-mul`
+and `no-short-forward-branch-i-mul` are the positive and negative directives for
+feature `short-forward-branch-i-mul` -- a specific kind of short forward branch
+optimization, respectively. Since it represents a specific kind of short forward
+branch optimization, the presence of this feature implies the presence of
+another _feature_ `short-forward-branch-opt`, which represents the general short forward
+branch optimization. This feature also follows a similar naming convention for its directives:
+`short-forward-branch-opt` for positive and `no-short-forward-branch-opt` for negative.
+Given this context, we have the following examples:
+
+`-mtune=rocket:short-forward-branch-i-mul` is a valid string, obviously. This results
+in both `short-forward-branch-i-mul` and `short-forward-branch-opt` being _enabled_
+among `rocket` 's existing performance characteristics.
+
+`-mtune=rocket:no-short-forward-branch-opt` is a valid string. This results in
+both `short-forward-branch-opt` and `short-forward-branch-i-mul` being _disabled_
+from `rocket` 's performance characteristics.
+
+`-mtune=rocket:short-forward-branch-i-mul,no-short-forward-branch-i-mul` is *invalid*.
+We do not allow these directives to "cancel out" each other.
+
+`-mtune=rocket:short-forward-branch-i-mul,short-forward-branch-i-mul` is *invalid*.
+We do not allow more than one occurrence of the same directive.
+
+`-mtune=rocket:short-forward-branch-i-mul,no-short-forward-branch-opt` is *invalid*.
+The `short-forward-branch-opt` implied by `short-forward-branch-i-mul` cannot
+be canceled out by `no-short-forward-branch-opt`.
+
+This _non-overlapping_ rule is a deliberate decision to simplify the logic and prevent
+this from becoming unmanageable in the future.
+
+It is worth noting that the actual tuning feature name is an implementation detail -- it
```

Comment:
This paragraph doesn't really explain what a "tuning feature name" is versus a "directive name". It assumes knowledge about LLVM/gcc implementation. I think including this paragraph just increases confusion.